pkg/util/tracing/crdbspan.go

```diff
@@ -566,12 +566,18 @@ func (s *crdbSpan) getLazyTagLocked(key string) (interface{}, bool) {
 
 // notifyEventListeners recursively notifies all the EventListeners registered
 // with this span and any ancestor spans in the Recording, of a StructuredEvent.
+//
+// If s has a parent, then we notify the parent of the StructuredEvent outside
+// the child (our current receiver) lock. This is as per the lock ordering
+// convention between parents and children.
 func (s *crdbSpan) notifyEventListeners(item Structured) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	if s.mu.recording.notifyParentOnStructuredEvent {
 		parent := s.mu.parent.Span.i.crdb
```

Comment:
Additionally, I don't think we should be holding on to `s.mu` mutex while calling `Notify` in the loop below. `Notify` could in turn be calling a method on the tracing span egs: `SetLazyTag` and that would cause a deadlock. I'm thinking of creating a copy of the `s.mu.eventListeners` slice while holding the mutex, and then operating on it outside of the span's mutex. 

Initially, I thought we could just operate on `s.mu.eventListeners` outside s.mu since we don't expect the slice to change given the only API to add event listeners is on creation. We could however race with the span being finished.